alonzo/impl/src/Cardano/Ledger/Alonzo/TxWitness.hs

```diff
@@ -0,0 +1,330 @@
+{-# LANGUAGE PatternSynonyms #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE StandaloneDeriving #-}
+{-# LANGUAGE StrictData #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+module Cardano.Ledger.Alonzo.TxWitness
+  ( RdmrPtr (..),
+    TxWitness (TxWitness, witsVKey, witsBoot, witsScript, witsData, witsRdmr),
+    EraIndependentScriptData,
+    ScriptDataHash,
+  )
+where
+
+import Cardano.Binary (FromCBOR (..), ToCBOR (..))
+import qualified Cardano.Crypto.Hash as Hash
+import Cardano.Ledger.Alonzo.Data (Data)
+import Cardano.Ledger.Alonzo.Scripts (Tag)
+import qualified Cardano.Ledger.Core as Core
+import Cardano.Ledger.Crypto (HASH)
+import Cardano.Ledger.Era (Era (Crypto))
+import Control.DeepSeq (NFData)
+import Data.Coders
+import Data.Map.Strict (Map)
+import Data.MemoBytes (Mem, MemoBytes (..), memoBytes)
+import Data.Set (Set)
+import qualified Data.Set as Set
+import Data.Typeable (Typeable)
+import Data.Word (Word64)
+import GHC.Generics
+import GHC.Records
+import NoThunks.Class (NoThunks)
+import Shelley.Spec.Ledger.Address.Bootstrap (BootstrapWitness)
+import Shelley.Spec.Ledger.Keys
+import Shelley.Spec.Ledger.Scripts (ScriptHash)
+import Shelley.Spec.Ledger.Serialization (mapFromCBOR, mapToCBOR)
+import Shelley.Spec.Ledger.TxBody (WitVKey)
+
+data RdmrPtr
+  = RdmrPtr
+      !Tag
+      {-# UNPACK #-} !Word64
+  deriving (Eq, Ord, Show, Generic)
+
+instance NoThunks RdmrPtr
+
+-- | Internal 'TxWitness' type, lacking serialised bytes.
+data TxWitnessRaw era = TxWitnessRaw
+  { _witsVKey :: Set (WitVKey 'Witness era),
+    _witsBoot :: Set (BootstrapWitness era),
+    _witsScriptData :: ScriptData era
+  }
+  deriving (Generic, Typeable)
+
+deriving stock instance
+  ( Era era,
+    Eq (Core.Script era)
+  ) =>
+  Eq (TxWitnessRaw era)
+
+deriving stock instance
+  (Era era, Show (Core.Script era)) =>
+  Show (TxWitnessRaw era)
+
+newtype TxWitness era = TxWitnessConstr (MemoBytes (TxWitnessRaw era))
+  deriving newtype (ToCBOR)
+
+deriving newtype instance (Era era, Eq (Core.Script era)) => Eq (TxWitness era)
+
+deriving newtype instance
+  (Era era, Show (Core.Script era)) =>
+  Show (TxWitness era)
+
+pattern TxWitness ::
+  (Era era, ToCBOR (Core.Script era)) =>
+  Set (WitVKey 'Witness era) ->
+  Set (BootstrapWitness era) ->
+  Map (ScriptHash era) (Core.Script era) ->
+  Set (Data era) ->
+  Map RdmrPtr (Data era) ->
+  TxWitness era
+pattern TxWitness
+  { witsVKey,
+    witsBoot,
+    witsScript,
+    witsData,
+    witsRdmr
+  } <-
+  TxWitnessConstr
+    ( Memo
+        ( TxWitnessRaw
+            witsVKey
+            witsBoot
+            ( ScriptDataConstr
+                ( Memo
+                    ( ScriptDataRaw
+                        witsScript
+                        witsData
+                        witsRdmr
+                      )
+                    _
+                  )
+              )
+          )
+        _
+      )
```

Comment:
Shouldn't we use the constructor patterns for ScriptData, rather than building it from whole cloth here?